pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	eventsTriggered := false
+	if !msg.Outbound {
+		// trigger the actionEvent if the message type is registered (if it's not outbound)
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. If called multiple times, the events will be sent to the
+// last channel. This works in conjunction with RegisterAutoExecute(). If this function is called after
+// RegisterAutoExecute(), the service will not execute the processing automatically and it'll wait for the callback. If
+// both are not set, then service will not handle the messages and throw an error.
+func (s *Service) RegisterActionEvent(ch chan<- event.DIDCommEvent) error {
+	s.lock.Lock()
+	s.actionEvent = ch
+	s.execute = true
+	s.lock.Unlock()
+
+	return nil
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (s *Service) UnregisterActionEvent() error {
+	return s.disableExecute()
+}
+
+// RegisterStreamEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The Callback is set to nil in the actionEvent message and service
+// won't be expecting any response from the consumers.
+func (s *Service) RegisterStreamEvent(ch chan<- dispatcher.DIDCommMsg) error {
```

Comment:
RegisterStatusEvent (or similar) might be the better corollary to RegisterActionEvent

Reply:
You may also want to match up the type names.

If you are calling the type DIDCommMsg then I would expect this method to be RegisterMsg[Foo].
Similarly with RegisterActionEvent returning DIDCommEvent then I would expect that method to be RegisterEvent[Foo].

Reply:
Updated.
1. RegisterActionEvent to [RegisterEvent](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R143)
2. RegisterStreamEvent to [RegisterMsg](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R160)